contents/handbook/engineering/clickhouse/clusters.mdx

```diff
@@ -0,0 +1,220 @@
+- Single 10TB GP3 EBS volume 
+
+
+We hit a problem with having smaller shards on EU that had a significant performance impact. We were running out of memory for larger queries which was also impacting our page cache hit rate.
+This was mainly due to limiting query size restrictions to protect other users of the cluster.
+We had two solutions for this. Increase the size of the nodes...meaning double the size for each instance x 16, very expensive. Or, we could setup a coordinator node.
+The coordinator node is a topology that allows us to effectively split the storage and compute tiers of the cluster into two pools of resources.
+We treat the current cluster of small nodes with many shards as the storage tier, they effectively are the mappers of the cluster and quickly fetch the data that we want for queries.
+We then send that relatively small data back to the coordinator and do the heavy lifting there which includes joins and aggregates.
+
+For the EU the coordinator node is:
+
+- 1 instance
+- `c7g.metal`
+- 64 vCPUs
+- metal Graviton3 CPU
+- 128 GiB RAM
+- Max Network 30 Gbps
+- Max EBS Throughput 2500 MB/s
+- 4 x 2.5TB GP3 EBS volumes
+    - JBOD configuration
+
+
+#### Coordinator schema
+
+The coordinator has distributed tables (`events`, `session_replay`) tables that point to the EU Prod cluster
+
+All non-sharded tables are replicated so that they are local to the coordinator.
+
+
+#### Coordinator future
+
+We should probably consider moving this to a `m7g.metal` if we hit any memory constraints with this, but so far we have not because of the dedicated nature of this node.
+
+We will also want to create new coodinators for multi-tennant workloads in the future. This will allow us to scale up and down easily over time, and even potentially throughout the day as the workload rises and falls.
+
+
+### Monitoring
+
+https://grafana.prod-eu.posthog.dev/d/vm-clickhouse-cluster-overview/clickhouse-cluster-overview
+
+
+## Problems
+
+The biggest pain points on our ClickHouse clusters is Disk Throughput. We still are using mutations too frequently.
+Every mutation rewrites large portions of our data on disk. This requires reading, and writing huge amounts of data which robs normal queries and inserts of resources.
+The best solution that we've found to support the current high utilization of mutations is to move to nodes that have local NVMe storage.
+This, along with RAID 10 far 2 configs provides us with roughly 1000 MB/s writes and 4000 MB/s reads at the same time on a node.
+This is  than anything we can get with any combination of EBS volumes alone (within reason $$$). A nice bonus on top of this is this does not impinge on the EBS throughput limits of the node.
+Meaning that on top of the baseline speed to NVMe disk we can tier out to EBS and have full instance EBS throughput available for that JBOD disk pack.
+
+Currently US is entirely on NVMe backed nodes. EU will need to be migrated to this setup as well.
+
+
+## Future work
```

Comment:
Are we going to keep both of the US and EU architectures as they are currently defined or at some point we will move to the coordinator topologu in US as well to get the benefits of compute / storage separation as well?

Reply:
Whatever we do in the long run - the goal is to have the same topology for all clusters